common/beerocks/bcl/include/bcl/network/socket.h

```diff
@@ -43,28 +43,79 @@ typedef unsigned __int64 ssize_t;
 
 class Socket {
 public:
-    explicit Socket(SOCKET s, long readTimeout = 1000);
-    Socket(SOCKET s, const std::string &peer_ip, int port_port, long readTimeout = 1000);
-    Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
-    virtual ~Socket();
+    // explicit Socket(SOCKET s, long readTimeout = 1000);
+    // Socket(SOCKET s, const std::string &peer_ip, int port_port, long readTimeout = 1000);
+    // Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
+    static Socket *socketFactory(SOCKET s, long readTimeout = 1000);
+    virtual ~Socket(){};
 
     enum SocketMode { SocketModeBlocking, SocketModeNonBlocking };
 
-    ssize_t getBytesReady();
-    ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
-                      bool isPeek = false, bool waitAll = false);
+    virtual ssize_t getBytesReady()                                      = 0;
+    virtual ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
+                              bool isPeek = false, bool waitAll = false) = 0;
     /**
      * @brief read and discard specified number of bytes
      *
      * @return error code if an error occured or number of discarded bytes otherwise
      */
     ssize_t discardBytes(size_t bytes_to_discard);
-    ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
-                       struct sockaddr_in addr_in = {});
+    virtual ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
+                               struct sockaddr_in addr_in = {}) = 0;
     ssize_t writeString(std::string msg)
     {
         return writeBytes((const uint8_t *)msg.c_str(), (size_t)msg.length());
     }
+    virtual bool setWriteTimeout(long msec) = 0;
+    virtual bool setReadTimeout(long msec)  = 0;
+    virtual void closeSocket()              = 0;
+    virtual bool isOpen()                   = 0;
+    SOCKET getSocketFd() { return m_socket; }
+    std::string getError() { return m_error; }
+    std::string getPeerIP() { return m_peer_ip; }
+    int getPeerPort() { return m_peer_port; }
+    std::string getUdsPath() { return m_uds_path; }
+    void setPeerMac(const std::string &mac) { m_peer_mac = mac; }
+    std::string getPeerMac() { return m_peer_mac; }
+
+    bool isAcceptedSocket() { return m_accepted_socket; }
+    void setIsServer() { m_is_server = true; }
+
+#ifndef IS_WINDOWS
+    virtual size_t getBytesWritePending() = 0;
+#endif
+
+protected:
+    friend class SocketServer;
+    friend class SocketSelect;
+    SOCKET m_socket = INVALID_SOCKET;
+    std::string m_error;
+    std::string m_peer_ip;
+    std::string m_uds_path;
+    int m_peer_port = 0;
+    std::string m_peer_mac;
+    bool m_accepted_socket  = false;
+    bool m_external_handler = false;
+    bool m_is_server        = false;
+
+private:
+    static int m_ref;
+};
+
+class DefaultSocket : public Socket {
```

Comment:
I didn't understand the commit message and the rationale of what you did here.
For a start, I don't understand the use of the word "mock" here.
AFAIK the meaning of it is "tease or laugh at in a scornful or contemptuous manner."
Can you please provide an explanation with an example to why or what it is helpful to create "DefaultSocket" class which is derived from "Socket" class. Why not to use "Socket" directly?

Reply:
The word "mock" is used here in a sence of "a fake object used to test the functionality of other object". Similar to "stub", but probably not the same.
This word is used in the name of GoogleMock framework. The basic idea is to create an abstract class and have two implementation derived from it: the real one and the fake "mocked" one. 
When testing a class that is using sockets we don't want it to fail due to socket issues because we want to test sockets separately. In this case socket user (e.g. beerocks_socket_thread) would recieve a mock to run a unit test.